### PR TITLE
[posix-host] use MultiFrameBuffer as SPI receive buffer

### PR DESCRIPTION
**Pull request by zhanglongxia**
This CL improves `MultiFrameBuffer` so that `MultiFrameBuffer` can reserve a space in front of the frame. SPI interface uses `MultiFrameBuffer` to store received frame directly, which avoids copying data from receive buffer to `MultiFrameBuffer`.

### CODE REVIEW

**Comment by abtink** [comment]
Thanks @zhanglongxia for this change and helping improve the `RadioSpinel` implementation.
I would like to propose a different approach.

The main idea for using a `MultiFrameBuffer` between HDLC and `RadioSpinel` is to avoid copying of frames from one buffer space to another. The model is as follows: 
- `RadioSpinel` defines and maintains a  `MultiFrameBuffer`.
- It gives this buffer to `HdclInterface` to use as buffer for its `Hdlc::Decoder`
-  As HDLC decoder decodes received data, it directly writes the decoded bytes into the `MultiFrameBuffer` space and once an entire frame is received/decoded it notifies next layer (`RadioSpinel`) to process the full frame.
- From the callback in `RadioSpinel` we check if we can process the received frame right then or not:
  - If we can process the new received frame (or if we don't care about the frame) after we are done, we call `DiscardFrame()` which frees up the space in `MultiFrameBuffer` for next received data.
  - If we cannot yet process the frame, we just `SaveFrame()` which keeps the frame in `MultiFrameBuffer` to be re-read layer.
- This way we never need to copy frame from one buffer to another. It always lives in same buffer space until it is processed and discarded.
- With the change from this PR we will always need to [copy the frame](https://github.com/openthread/openthread/pull/4577/files#diff-8e57fb9200bcbf4df1430f45821ceeddR359) in both SPI and UART cases from `RadioSpinel` (this is contrast to existing code which would only require copying of frame under SPI interface).

We currently use `RadioSpinel` under posix platform and we can argue that copying frames from buffer to buffer should be fine under posix but we may want to use the same code/model on other more resource restricted platforms (say `RadioSpinel` or `RCP host` code on RTOS style platforms).

Here is my suggestion for a different approach:

- The idea is if we can somehow chnage SPI interface code such that it would behave similar to `Hdlc::Decoder` and directly uses the `MultiFrameBuffer` and writes the received frame into the same buffer (to avoid the need for copying under SPI).
- One difference I see between HDLC and SPI frame is that SPI frame has some additional SPI specific header (and also may require some extra bytes for alignments).
  - We can make it work by adding some mechanism to inform next layer to skip some portion of frame (e.g.., provide an `skipLength` to show how many bytes to skip from received data frame to get to actual start of frame (the actual content))
  - We can consider adding such a mechanism in the `MultiFrameBuffer` itself or maybe a different sub-class of `FrameWritePointer`.

 Thoughts?

**Comment by zhanglongxia** [comment]
@abtink Thanks for your detailed explanation about the usage of the `MultiFrameBuffer` and provide suggestions.

 When we are  sending a spinel frame, we are using the  function `SendFrame(const uint8_t *aFrame, uint16_t aLength)` to send frame. But when we are receiving a frame we are using a `MultiFrameBuffer` to store the received frame.  I think the `SendFrame(const uint8_t *aFrame, uint16_t aLength)` and `HandleReceivedFrame()` function should be consistent. 

Besides, we want to compile` SpiInterface` and `HdlcInterface` into two different libraries later, and posix-host loads different libraries based on different parameters. If we are using `MultiFrameBuffer`, It is hard for us to use 'MultiFrameBuffer' to call the libraries.

For the `memory copy ` issue, we can directly process the received spinel frame. If the spinel frame doesn't need to be saved, it won't be copied into `MultiFrameBuffer`. If the spinel frame needs to be saved, just copy it into the 'MultiFrameBuffer'. This method avoid copying all received frames into `MultiFrameBuffer`, only the out-of-order packets will be copied into  `MultiFrameBuffer`. The number of out-of-order packets will not be large.

Comments?

**Comment by abtink** [comment]
> When we are sending a spinel frame, we are using the function SendFrame(const uint8_t *aFrame, uint16_t aLength) to send frame. But when we are receiving a frame we are using a MultiFrameBuffer to store the received frame. I think the SendFrame(const uint8_t *aFrame, uint16_t aLength) and HandleReceivedFrame() function should be consistent.

I would say having similar looking APIs for send/receive would be good but I would definitely favor functionality over it. We have other places where send/receive APIs do not mirror each other like in `SubMac` or `Mac` classes.

> Besides, we want to compile `SpiInterface` and `HdlcInterface` into two different libraries later, and posix-host loads different libraries based on different parameters. If we are using MultiFrameBuffer, It is hard for us to use 'MultiFrameBuffer' to call the libraries.

We can consider moving the buffer implementation into its own library. The `MultiFrameBuffer` was originally added in `hdlc.cpp` (since it was its only user), but if we expand its use we can move it into its own module. 

> For the memory copy issue, we can directly process the received spinel frame. If the spinel frame doesn't need to be saved, it won't be copied into MultiFrameBuffer. If the spinel frame needs to be saved, just copy it into the 'MultiFrameBuffer'. This method avoid copying all received frames into MultiFrameBuffer, only the out-of-order packets will be copied into MultiFrameBuffer. The number of out-of-order packets will not be large.

In current uart/hdlc model we would never need to copy any received frame. The frame is directly written into the same buffer space (by hdlc decoder) and either processed (and then buffer space is freed) or placed in a queue (as is without copying it). 

I am not sure if we can change the radio-spinel design to not require queuing of some received spinel frame. We should never get out of order frames (in terms of spinel TIDs) over spinel, but it is common to receive async data before a response (e.g., receiving data frames over radio which could be emitted over spinel before a spinel command response). 

I would suggest if we can write the code in an optimized way that avoids the need for extra copying and would allow us to use the same implementation under posix and later RTOS, it would be very valuable.

We can consider changing SPI code to also use `MultiFrameBuffer` directly (i.e. write the received frame over SPI directly into the buffer provided by `MultiFrameBuffer`). We can consider extending  `MultiFrameBuffer` to allow a  `SkipLength` to be specified for each frame by the receiver of frame (i.e., HDLC decoder or SPI receiver). The SPI driver would then set the skip length such that we would skip the SPI header portion when reading/parsing the frame in next layer `RadioSpinel.

Thoughts?

**Comment by zhanglongxia** [comment]
@abtink  Thanks for your suggestion. I will update the code according to your suggestion.

File: src/lib/hdlc/hdlc.hpp [hunk 1]
```diff
@@ -265,7 +333,7 @@ template <uint16_t kSize> class MultiFrameBuffer : public FrameWritePointer
      */
     void DiscardFrame(void)
     {
-        mWritePointer    = mWriteFrameStart + kHeaderSize;
+        mWritePointer    = GetFrame();
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/lib/hdlc/hdlc.hpp]: Updated.

File: src/posix/platform/spi_interface.cpp [hunk 2]
```diff
@@ -266,18 +266,20 @@ void SpiInterface::TrigerReset(void)
     otLogNotePlat("Triggered hardware reset");
 }
 
-uint8_t *SpiInterface::GetRealRxFrameStart(void)
+uint8_t *SpiInterface::GetRealRxFrameStart(uint8_t *aSpiRxFrameBuffer, uint8_t aAlignAllowance, uint16_t *aSkipLength)
```
[the remark this answers is not in the document]
Comment (zhanglongxia) [reply src/posix/platform/spi_interface.cpp]: Updated.

**Pull request merged**